mypy/ipc.py

```diff
@@ -47,36 +47,66 @@ class IPCBase:
 
     connection = None  # type: _IPCHandle
 
-    def __init__(self, name: str) -> None:
+    def __init__(self, name: str, timeout: Optional[float]) -> None:
         self.READ_SIZE = 100000
         self.name = name
+        if sys.platform == 'win32':
+            # Windows timeouts are in miliseconds
+            self.timeout = int(timeout * 1000) if timeout is not None else None
+        else:
+            self.timeout = timeout
 
     def read(self) -> bytes:
         """Read bytes from an IPC connection until its empty."""
         bdata = bytearray()
-        while True:
-            if sys.platform == 'win32':
-                more, _ = _winapi.ReadFile(self.connection, self.READ_SIZE)
-            else:
+        if sys.platform == 'win32':
+            while True:
+                ov, err = _winapi.ReadFile(self.connection, self.READ_SIZE, overlapped=True)
+                # TODO: remove once typeshed supports Literal types
+                assert isinstance(ov, _winapi.Overlapped)
+                assert isinstance(err, int)
+                try:
+                    if err == _winapi.ERROR_IO_PENDING:
+                        timeout = self.timeout if self.timeout else _winapi.INFINITE
+                        res = _winapi.WaitForSingleObject(ov.event, timeout)
+                        assert res == _winapi.WAIT_OBJECT_0
+                except BaseException:
+                    ov.cancel()
+                    raise
+                _, err = ov.GetOverlappedResult(True)
```

Comment:
I suppose this should be reached when `err` is zero as well, indicating that somehow the `ReadFile` operation returned an instant result? So far so good. But what if `err` has some *other* value? Will `GetOverlappedResult` just pass that through? Or should you check for that?

Reply:
I took a look at the [source](https://github.com/python/cpython/blob/master/Modules/_winapi.c#L202) and the possible values the err value can take (other values raise exceptions):
```
_winapi.ERROR_SUCCESS
_winapi.ERROR_MORE_DATA
_winapi.ERROR_OPERATION_ABORTED
_winapi.ERROR_IO_INCOMPLETE
```
`_winapi.ERROR_IO_INCOMPLETE` can't happen in blocking mode (which we are). More data means we should read again, so I guess I just need to handle operation aborted (it seems CPython does not, for some reason).

Reply:
OK, then maybe the code in the try/except should do something like
```
if err != 0:
    assert err == _winapi.ERROR_IO_PENDING
    # rest of the code
```